src/riak_kv_vnode.erl

```diff
@@ -2322,102 +2346,286 @@ new_md_cache(VId) ->
+    MaxTime = non_neg_env(riak_kv, counter_lease_timeout, ?DEFAULT_CNTR_LEASE_TO),
+    {MaxErrors, MaxTime}.
+
+%% @private we don't want to crash riak because of a dodgy config
+%% value, and by this point, cuttlefish be praised, we should have
+%% sane values. However, if not, ignore negative values for timeout
+%% and max errors, use the defaults, and log the insanity. Note this
+%% expects the macro configured defaults to be positive integers!
+-spec non_neg_env(atom(), atom(), pos_integer()) -> pos_integer().
+non_neg_env(App, EnvVar, Default) when is_integer(Default),
+                                       Default > 0 ->
+    case app_helper:get_env(App, EnvVar, Default) of
+        N when is_integer(N),
+               N > 0 ->
+            N;
+        X ->
+            lager:warning("Non-integer/Negative integer ~p for vnode counter config ~p."
+                          " Using default ~p",
+                          [X, EnvVar, Default]),
+            Default
+    end.
+
+%% @private to keep put_merge/6 side effect free (as it is exported
+%% for testing) introspect the state and local/incoming objects and
+%% return the actor ID for a put, and possibly updated state. NOTE
+%% side effects, in that the vnode status on disk can change.
+%% Why might we need a new key epoch?
+%% 1. local not found (handled in prepare_put/3
+%% 2. local found, but never acted on this key before (or don't remember it!)
+%% 3. local found, local acted, but incoming has greater count or actor epoch
+%%    This one is tricky, since it indicates some byzantine failure somewhere.
+-spec maybe_new_key_epoch(boolean(), #state{},
+                          riak_object:riak_object(),
+                          riak_object:riak_object()) ->
+                                 {binary(), #state{}}.
+maybe_new_key_epoch(false, State, _, _) ->
+    %% Never add a new key epoch when not coordinating
+    %% @TODO (rdb) need to mark actor as dirty though.
+    {State#state.vnodeid, State};
+maybe_new_key_epoch(true, State, LocalObj, IncomingObj) ->
+    #state{vnodeid=VId} = State,
+    %% @TODO (rdb) maybe optimise since highly likey both objects
+    %% share the majority of actors, maybe a single umerged list of
+    %% actors, somehow tagged by local | incoming?
+    case highest_actor(VId, LocalObj) of
+        {undefined, 0, 0} -> %% Not present locally
+            %% Never acted on this object before, new epoch.
+            new_key_epoch(State);
+        {LocalId, LocalEpoch, LocalCntr} -> %% Present locally
+            case highest_actor(VId, IncomingObj) of
+                {_InId, InEpoch, InCntr} when InEpoch > LocalEpoch;
+                                              InCntr > LocalCntr ->
+                    %% In coming actor-epoch or counter greater than
+                    %% local, some byzantine failure, new epoch.
+                    %%
+                    %% @TODO (rdb) do we log/stat here?
```

Comment:
I’m not sure about “Late" and "Old message"?

The vnode has sent a lease message to the manager. It has yet to receive an answer so it enters this blocking receive. What is the old message? Do you mean if we timeout and crash the vnode? There is no old message as the message queue dies with the crashed vnode.

Is this the sequence you’re thinking of that leads to there being an “old message” for the lease

1. vnode asks for a lease
2. manager sends a lease message
3. manager crashes
4. vnode in blocking lease receive gets the exit message first (‘cos of the order of receives in statement)
5. vnode starts a new manager and requests a lease
6. vnode pulls the original message (pre crash)

Is that step 6 the “old message?”

I don’t think it can happen. The `Pid` in the receive is the new managers `Pid`, the old one will never match. Even if it would I think it is safe. The vnode is getting a new lease! Yay! It can keep working. The lease it gets is smaller than that on disk, yay! Invariant preserved. I assume then there is a step 7

7. vnode gets new lease message

All is still OK. We just have a larger lease/longer until we re-lease. If the vnode crashes it just wasted some integers.

Does that seem OK to you? (This is the first time I have thought it through, so I think it was a valid question, if I answered the question you asked, that is.)

I don’t think there is any way for the vnode to request a new lease before it receives the old lease, and Erlang delivers messages between processes on the same node in order, so they’ll be received in order, meaning the lease always increases. Happy to mumble about it in the morning to be sure, though.

> On 13 Jan 2015, at 20:55, Andrew J. Stone <notifications@github.com> wrote:
> 
> In src/riak_kv_vnode.erl:
> 
> > +    #state{vnodeid=VId} = State,
> > +    %% @TODO (rdb) maybe optimise since highly likey both objects
> > +    %% share the majority of actors, maybe a single umerged list of
> > +    %% actors, somehow tagged by local | incoming?
> > +    case highest_actor(VId, LocalObj) of
> > +        {undefined, 0, 0} -> %% Not present locally
> > +            %% Never acted on this object before, new epoch.
> > +            new_key_epoch(State);
> > +        {LocalId, LocalEpoch, LocalCntr} -> %% Present locally
> > +            case highest_actor(VId, IncomingObj) of
> > +                {_InId, InEpoch, InCntr} when InEpoch > LocalEpoch;
> > +                                              InCntr > LocalCntr ->
> > +                    %% In coming actor-epoch or counter greater than
> > +                    %% local, some byzantine failure, new epoch.
> > +                    %%
> > +                    %% @TODO (rdb) do we log/stat here?
> 
> I would add at least a warning log here. Either there is a bug in the backend code or the disk is losing things.
> 
> —
> Reply to this email directly or view it on GitHub.
>